pkg/datagatherer/k8s/dynamic_test.go

```diff
@@ -1009,3 +1086,200 @@ func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+					"key1": "value1",
+					"key2": "value2",
+				},
+			},
+		},
+		expectObj: map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"name": "foo",
+				"annotations": map[string]interface{}{
+					"key1": "value1",
+					"key2": "value2",
+				},
+			},
+		},
+	}))
+
+	// The "leaf" field is the field that is at the end of the path. For
+	// example, "annotations" is the leaf field in metadata.annotations.
+	t.Run("works when the leaf field is not found", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath:    []string{"metadata", "annotations"},
+		givenExclude: []string{},
+
+		givenObj:  map[string]interface{}{"metadata": map[string]interface{}{}},
+		expectObj: map[string]interface{}{"metadata": map[string]interface{}{}},
+	}))
+
+	t.Run("works when the leaf field is nil", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath:    []string{"metadata", "annotations"},
+		givenExclude: []string{},
+		givenObj: map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"name":        "foo",
+				"annotations": nil,
+			},
+		},
+		expectObj: map[string]interface{}{"metadata": map[string]interface{}{"name": "foo"}},
+	}))
+
+	t.Run("works when leaf field is unexpectedly not nil and not a known map", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath: []string{"metadata", "annotations"},
+		givenObj:  map[string]interface{}{"metadata": map[string]interface{}{"annotations": 42}},
+		expectObj: map[string]interface{}{"metadata": map[string]interface{}{"annotations": 42}},
+	}))
+
+	// The "intermediate" field is the field that is not at the end of the path.
+	// For example, "metadata" is the intermediate field in
+	// metadata.annotations.
+	t.Run("works when the intermediate field doesn't exist", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath: []string{"metadata", "annotations"},
+		givenObj:  map[string]interface{}{},
+		expectObj: map[string]interface{}{},
+	}))
+
+	t.Run("works when the intermediate field is nil", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath: []string{"metadata", "annotations"},
+		givenObj:  map[string]interface{}{"metadata": nil},
+	}))
+
+	t.Run("works when the intermediate field is unexpectedly not nil and not a map", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
+		givenPath: []string{"metadata", "annotations"},
+		givenObj:  map[string]interface{}{"metadata": 42},
+		expectObj: map[string]interface{}{"metadata": 42},
+	}))
+}
+
+type tc_RemoveUnstructuredKeys struct {
+	givenExclude []string
+	givenObj     map[string]interface{}
+	givenPath    []string
+	expectObj    map[string]interface{}
+}
+
+func run_TestRemoveUnstructuredKeys(tc tc_RemoveUnstructuredKeys) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Helper()
+		RemoveUnstructuredKeys(toRegexps(tc.givenExclude), &unstructured.Unstructured{Object: tc.givenObj}, tc.givenPath...)
```

Comment:
```suggestion
		RemoveUnstructuredKeys(toRegexps(tc.givenExclude), &unstructured.Unstructured{Object: tc.givenObj}, tc.givenPath...)
		assert.Equal(t, tc.expectObj, tc.givenObj)
```

Reply:
After adding the missing assertion the test cases needed some adjustment.

```diff
diff --git a/pkg/datagatherer/k8s/dynamic_test.go b/pkg/datagatherer/k8s/dynamic_test.go
index 8002a15..fcf0071 100644
--- a/pkg/datagatherer/k8s/dynamic_test.go
+++ b/pkg/datagatherer/k8s/dynamic_test.go
@@ -1103,6 +1103,9 @@ func TestRemoveUnstructuredKeys(t *testing.T) {
                expectObj: map[string]interface{}{
                        "metadata": map[string]interface{}{
                                "name": "foo",
+                               "annotations": map[string]interface{}{
+                                       "key2": "value2",
+                               },
                        },
                },
        }))
@@ -1121,7 +1124,8 @@ func TestRemoveUnstructuredKeys(t *testing.T) {
                },
                expectObj: map[string]interface{}{
                        "metadata": map[string]interface{}{
-                               "name": "foo",
+                               "name":        "foo",
+                               "annotations": map[string]interface{}{},
                        },
                },
        }))
@@ -1140,7 +1144,8 @@ func TestRemoveUnstructuredKeys(t *testing.T) {
                },
                expectObj: map[string]interface{}{
                        "metadata": map[string]interface{}{
-                               "name": "foo",
+                               "name":        "foo",
+                               "annotations": map[string]interface{}{},
                        },
                },
        }))
@@ -1187,7 +1192,12 @@ func TestRemoveUnstructuredKeys(t *testing.T) {
                                "annotations": nil,
                        },
                },
-               expectObj: map[string]interface{}{"metadata": map[string]interface{}{"name": "foo"}},
+               expectObj: map[string]interface{}{
+                       "metadata": map[string]interface{}{
+                               "name":        "foo",
+                               "annotations": nil,
+                       },
+               },
        }))

        t.Run("works when leaf field is unexpectedly not nil and not a known map", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
@@ -1208,6 +1218,7 @@ func TestRemoveUnstructuredKeys(t *testing.T) {
        t.Run("works when the intermediate field is nil", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
                givenPath: []string{"metadata", "annotations"},
                givenObj:  map[string]interface{}{"metadata": nil},
+               expectObj: map[string]interface{}{"metadata": nil},
        }))

        t.Run("works when the intermediate field is unexpectedly not nil and not a map", run_TestRemoveUnstructuredKeys(tc_RemoveUnstructuredKeys{
@@ -1228,6 +1239,7 @@ func run_TestRemoveUnstructuredKeys(tc tc_RemoveUnstructuredKeys) func(*testing.
        return func(t *testing.T) {
                t.Helper()
                RemoveUnstructuredKeys(toRegexps(tc.givenExclude), &unstructured.Unstructured{Object: tc.givenObj}, tc.givenPath...)
+               assert.Equal(t, tc.expectObj, tc.givenObj)
        }
 }
```

Reply:
Thanks for noticing this... I rushed this part of the testing, it clearly shows. Thanks for proposing a fix!

I've adjusted the tests.